pkg/snet/stun_handler.go

```diff
@@ -0,0 +1,388 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package snet
+
+import (
+	"context"
+	"errors"
+	"net"
+	"net/netip"
+	"os"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/scionproto/scion/pkg/stun"
+)
+
+const timeoutDuration = 5 * time.Minute
+
+// stunHandler is a wrapper around net.UDPConn that handles STUN requests.
+type stunHandler struct {
+	*net.UDPConn
+	recvChan       chan bufferedPacket
+	maxQueuedBytes int64
+	mutex          sync.Mutex
+
+	// the following fields are protected by mutex
+	queuedBytes          int64
+	stunChans            map[stun.TxID]chan stunResponse
+	mappings             map[*net.UDPAddr]*natMapping
+	retransmissionTimers map[*net.UDPAddr]*retransmissionTimer
+	pendingRequests      map[*net.UDPAddr]bool
+	writeDeadline        time.Time
+	readDeadline         time.Time
+	cond                 *sync.Cond // condition variable for pending STUN requests
+}
+
+type bufferedPacket struct {
+	data []byte
+	addr net.Addr
+}
+
+type stunResponse struct {
+	mappedAddr netip.AddrPort
+	err        error
+}
+
+func newSTUNHandler(conn *net.UDPConn) (*stunHandler, error) {
+	// Get the receive buffer size
+	fd, err := conn.File()
+	if err != nil {
+		return nil, err
+	}
+	defer fd.Close()
+	rcvBufSize, err := syscall.GetsockoptInt(int(fd.Fd()), syscall.SOL_SOCKET, syscall.SO_RCVBUF)
+	if err != nil {
+		return nil, err
+	}
+	maxNumPacket := rcvBufSize / 64 // assuming lower bound of per packet metadata of 64 bytes
+	if maxNumPacket < 10 {
+		maxNumPacket = 10
+	}
+
+	handler := &stunHandler{
+		UDPConn:              conn,
+		recvChan:             make(chan bufferedPacket, maxNumPacket),
+		maxQueuedBytes:       int64(rcvBufSize),
+		stunChans:            make(map[stun.TxID]chan stunResponse),
+		mappings:             make(map[*net.UDPAddr]*natMapping),
+		retransmissionTimers: make(map[*net.UDPAddr]*retransmissionTimer),
+		pendingRequests:      make(map[*net.UDPAddr]bool),
+	}
+	handler.cond = sync.NewCond(&handler.mutex)
+
+	// background goroutine to continuously read from the underlying UDP connection and filter out
+	// STUN packets
+	go func() {
+		buf := make([]byte, 1500)
+		for {
+			n, addr, err := handler.UDPConn.ReadFrom(buf)
+			if err != nil {
+				if errors.Is(err, net.ErrClosed) ||
+					errors.Is(err, syscall.EBADF) { // bad file descriptor (connection closed)
+					close(handler.recvChan)
+					return
+				}
+				continue
+			}
+
+			data := make([]byte, n)
+			copy(data, buf[:n])
+
+			if stun.Is(data) {
+				respTxID, mappedAddr, err := stun.ParseResponse(data)
+				if err != nil {
+					continue
+				}
+				handler.mutex.Lock()
+				ch, ok := handler.stunChans[respTxID]
+				handler.mutex.Unlock()
+				if ok {
+					select {
+					case ch <- stunResponse{mappedAddr: mappedAddr, err: err}:
+					default:
+					}
+				}
+			} else {
+				handler.queuePacket(bufferedPacket{data: data, addr: addr})
```

Comment:
Either make the first branch of the if statement a method too or inline this branch. (I'd probably just inline `queuePacket`)